addons/l10n_in/models/company.py

```diff
@@ -25,3 +48,23 @@ def _update_l10n_in_fiscal_position(self):
             ChartTemplate = self.env['account.chart.template'].with_company(company)
             fiscal_position_data = ChartTemplate._get_in_account_fiscal_position()
             ChartTemplate._load_data({'account.fiscal.position': fiscal_position_data})
+
+    @api.onchange('vat')
+    def onchange_vat(self):
+        if self.vat and self.check_vat_in(self.vat):
+            state_id = self.env['res.country.state'].search([('l10n_in_tin', '=', self.vat[:2])], limit=1)
+            if state_id:
+                self.state_id = state_id
+
+    def check_vat_in(self, vat):
+        # reference from https://www.gstzen.in/a/format-of-a-gst-number-gstin.html
+        if vat and len(vat) == 15:
```

Comment:
You can remove the check for the existence of the `vat` variable if you set a default and it makes the method safer :+1: 

```suggestion
    def check_vat_in(self, vat=""):
        # reference from https://www.gstzen.in/a/format-of-a-gst-number-gstin.html
        if len(vat) == 15:
```